### PR TITLE
feat: external annotator for OpenFGA files

### PR DESCRIPTION
**Pull request by d-jeffery**
## Description

[openfgaintellijplugin-0.1.1.zip](https://github.com/user-attachments/files/15516376/openfgaintellijplugin-0.1.1.zip)

## References

- closes https://github.com/openfga/intellij-plugin/issues/17

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -15,10 +15,13 @@ repositories {
 }
 
 dependencies {
+    implementation("org.antlr:antlr4:4.13.1")
     implementation("dev.openfga:openfga-sdk:0.4.2")
     implementation("org.dmfs:oauth2-essentials:0.22.1")
     implementation("org.dmfs:httpurlconnection-executor:1.22.1")
 
+    implementation(files("libs/java-0.0.1.jar"))
```
Comment (d-jeffery) [inline build.gradle.kts:23]: Needs to be replaced with published library

File: src/main/java/dev/openfga/intellijplugin/language/OpenFGAStoreAnnotator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                    .range(new TextRange(offsetStart, offsetEnd))
+                    .create();
+        }
+    }
+
+    private static int getOffsetFromRange(@NotNull final String doc, int line, int character, int offset) {
+        final String[] lines = doc.split("\n");
+
+        for (int i = 0; i < line; i++) {
+            final String replacementLine = lines[i].replaceFirst("^  ", "");
```
Comment (d-jeffery) [inline src/main/java/dev/openfga/intellijplugin/language/OpenFGAStoreAnnotator.java:82]: This is very brittle - need to tackle cases when 0, 1 spaces, or tabs

File: build.gradle.kts (hunk 1 above)
Comment (rhamzeh) [reply build.gradle.kts]: I think, once published language will be:
```
    implementation("dev.openfga:openfga-language:0.2.0")
```

but for now, can you add that in a comment? That this cannot be built without manually building language and providing the jar file.

And let's merge and tag

**Comment by rhamzeh** [comment]
What do we need to provide so that this value is filled?
Screenshot 2024-05-31 at 8 16 02 AM

**Comment by d-jeffery** [comment]
> What do we need to provide so that this value is filled? 

https://plugins.jetbrains.com/docs/intellij/plugin-configuration-file.html#idea-plugin__change-notes

Will update.

File: build.gradle.kts (hunk 1 above)
Comment (d-jeffery) [reply build.gradle.kts]: Comment added.

File: build.gradle.kts [hunk 3]
```diff
@@ -7,17 +7,22 @@ plugins {
...
 }
 
 dependencies {
+    implementation("org.antlr:antlr4:4.13.1")
     implementation("dev.openfga:openfga-sdk:0.4.2")
     implementation("org.dmfs:oauth2-essentials:0.22.1")
     implementation("org.dmfs:httpurlconnection-executor:1.22.1")
+    implementation("org.apache.commons:commons-lang3:3.14.0")
+
+    // Plugin cannot be built without manually building language and providing the jar file
```
Comment (rhamzeh) [inline build.gradle.kts:24]:
```suggestion
    // Until, https://github.com/openfga/language/pkg/java is published, the plugin cannot be built without manually building language and providing the jar file
```

**Review by rhamzeh — APPROVED** [review]

**Pull request merged**